coalib/files/Fileproxy.py

```diff
@@ -0,0 +1,97 @@
+from coala_utils.decorators import generate_eq
+
+
+@generate_eq("filename")
+class Fileproxy:
+    """
+    The file proxy is a wrapper object to wrap every file. The file proxy is an
+    object that contains properties about the file currently processed (such as
+    the filename). The fileproxy object also protects your file from being
+    modified by a bear. It contains:
+
+        * The filename, this is the starting point for every other attribute.
+        * The content of the file as a string.
+        * The content of the file, line-splitted.
+
+    Note: Other properties can be included later as and when required.
+
+    The equality of the object is checked just based on its filename.
+
+    To initialize a Fileproxy object for a file:
+
+    >>> fp = Fileproxy("./tests/files/FileproxyTestFiles/testfile.txt")
+
+    Now we can get the filename:
+
+    >>> fp.filename
+    './tests/files/FileproxyTestFiles/testfile.txt'
+
+    The number of lines in the file:
+
+    >>> len(fp)
+    1
+
+    Fetch a specific line in the file:
+
+    >>> fp[0]
+    'Old line\\n'
+
+    Iterate through the lines of the file:
+
+    >>> for line in fp:
+    ...     print(line)
+    Old line
+    <BLANKLINE>
+
+    Get contents of the file as a string:
+
+    >>> str(fp)
+    'Old line\\n'
+
+    Get all lines of the file as a list:
+    >>> list(fp)
+    ['Old line\\n']
+    """
+
+    def __init__(self, filename):
+        """
+        Constructs a new fileproxy object.
+
+        :param filename: The name of the file to load.
+        """
+        self.filename = filename
+        with open(self.filename, encoding="utf-8") as filehandle:
+            self.__content = filehandle.read()
+        self.__lines = tuple(self.__content.splitlines(True))
```

Comment:
So we actually did this to increase the speed of coala. As the fileproxy will be used by multiple bears hence it will have the content and the lines ready to go whenever a bear uses it. Avoiding to load it multiple times. This Fileproxy will also be available across sections and hence it will save time their too.